mitiq/pec/pec.py

```diff
@@ -0,0 +1,106 @@
+       circuits (through the "executor" function provided by the user);
+    3. Estimating the ideal expectation value from a suitable linear
+       combination of the noisy ones.
+
+    Args:
+        circuit: The input circuit to execute with error-mitigation.
+        executor: A function which executes a circuit and returns an
+            expectation value.
+        deco_dict: The decomposition dictionary containing the quasi-
+            probability representation of the ideal operations (those
+            which are part of the input circuit).
+        num_samples: The number of noisy circuits to be sampled for PEC.
+            If equal to None, it is deduced from the amount of "negativity"
+            of the quasi-probability representation of the input circuit.
+            Note: the latter feature is not yet implemented and num_samples
+            is just set to 1000 if not specified.
+        full_output: If False only the average PEC value is returned.
+            If True the associated error (standard deviation) is returned too.
+
+    Returns:
+        pec_value: The PEC estimate of the ideal expectation value associated
+            to the input circuit.
+        pec_std: The standard deviation of the PEC samples, i.e., the square
+            root of the mean squared deviation (from pec_value).
+            After dividing this quantity by sqrt(num_samples), one obtains
+            an estimate of the error between 'pec_value' and the actual
+            ideal expectation value.
+
+    .. [Temme2017] : Kristan Temme, Sergey Bravyi, Jay M. Gambetta,
+        "Error mitigation for short-depth quantum circuits,"
+        *Phys. Rev. Lett.* **119**, 180509 (2017),
+        (https://arxiv.org/abs/1612.02058).
+
+    .. [Endo2018] : Suguru Endo, Simon C. Benjamin, Ying Li,
+        "Practical Quantum Error Mitigation for Near-Future Applications"
+        *Phys. Rev. **X 8**, 031027 (2018),
+        (https://arxiv.org/abs/1712.09271).
+
+    .. [Takagi2020] : Ryuji Takagi,
+        "Optimal resource cost for error mitigation,"
+        (https://arxiv.org/abs/2006.12509).
+    """
+
+    # TODO: maybe deduce num_to_sample from the decomposition "negativity"
+    if not num_samples:
+        num_samples = 1000
+
+    sampled_circuits = []
+    signs = []
+
+    for _ in range(num_samples):
+        # Note: the norm is the same for each sample.
+        sampled_circuit, sign, norm = sample_circuit(circuit, deco_dict)
+        sampled_circuits.append(sampled_circuit)
+        signs.append(sign)
+
+    # TODO: add support for batched executors
+    # Execute all the circuits
+    exp_values = [executor(circ) for circ in sampled_circuits]
+
+    # Evaluate unbiased estimators [Temme2017], [Endo2018], [Takagi2020]
+    unbiased_estimators = [norm * s * val for s, val in zip(signs, exp_values)]
+
+    if full_output:
+        return np.average(unbiased_estimators), np.std(unbiased_estimators)
```

Comment:
I used the `full_output` options for 3 reasons which are however not very strong:

1. For symmetry with `execute_with_zne`.
2. Because `execute_with_pec` is the entry point for a non-expert user which could be confused by additional outputs.
3. Because we may want to return many results in the future. (In some way to be discussed #414).

I created new issue  #414  about this and in particular about how to report raw results of a PEC process.